aspnetcore/blazor/advanced-scenarios.md

```diff
@@ -0,0 +1,163 @@
+ms.author: riande
+ms.custom: mvc
+ms.date: 02/12/2020
+no-loc: [Blazor, SignalR]
+uid: blazor/advanced-scenarios
+---
+# ASP.NET Core Blazor advanced scenarios
+
+By [Luke Latham](https://github.com/guardrex) and [Daniel Roth](https://github.com/danroth27)
+
+## Manual RenderTreeBuilder logic
+
+`Microsoft.AspNetCore.Components.Rendering.RenderTreeBuilder` provides methods for manipulating components and elements, including building components manually in C# code.
+
+> [!NOTE]
+> Use of `RenderTreeBuilder` to create components is an advanced scenario. A malformed component (for example, an unclosed markup tag) can result in undefined behavior.
+
+Consider the following `PetDetails` component, which can be manually built into another component:
+
+```razor
+<h2>Pet Details Component</h2>
+
+<p>@PetDetailsQuote</p>
+
+@code
+{
+    [Parameter]
+    public string PetDetailsQuote { get; set; }
+}
+```
+
+In the following example, the loop in the `CreateComponent` method generates three `PetDetails` components. When calling `RenderTreeBuilder` methods to create the components (`OpenComponent` and `AddAttribute`), sequence numbers are source code line numbers. The Blazor difference algorithm relies on the sequence numbers corresponding to distinct lines of code, not distinct call invocations. When creating a component with `RenderTreeBuilder` methods, hardcode the arguments for sequence numbers. **Using a calculation or counter to generate the sequence number can lead to poor performance.** For more information, see the [Sequence numbers relate to code line numbers and not execution order](#sequence-numbers-relate-to-code-line-numbers-and-not-execution-order) section.
+
+`BuiltContent` component:
+
+```razor
+@page "/BuiltContent"
+
+<h1>Build a component</h1>
+
+@CustomRender
+
+<button type="button" @onclick="RenderComponent">
+    Create three Pet Details components
+</button>
+
+@code {
+    private RenderFragment CustomRender { get; set; }
+    
+    private RenderFragment CreateComponent() => builder =>
+    {
+        for (var i = 0; i < 3; i++) 
+        {
+            builder.OpenComponent(0, typeof(PetDetails));
+            builder.AddAttribute(1, "PetDetailsQuote", "Someone's best friend!");
+            builder.CloseComponent();
+        }
+    };    
+    
+    private void RenderComponent()
+    {
+        CustomRender = CreateComponent();
+    }
+}
+```
+
+> [!WARNING]
+> The types in `Microsoft.AspNetCore.Components.RenderTree` allow processing of the *results* of rendering operations. These are internal details of the Blazor framework implementation. These types should be considered *unstable* and subject to change in future releases.
+
+### Sequence numbers relate to code line numbers and not execution order
+
+Blazor `.razor` files are always compiled. This is potentially a great advantage for `.razor` because the compile step can be used to inject information that improve app performance at runtime.
```

Comment:
```suggestion
Blazor *.razor* files are always compiled. This is potentially a great advantage for *.razor* because the compile step can be used to inject information that improves app performance at runtime.
```

> This is potentially a great advantage...

A great advantage over what? As a reader, I'd want that information.